include/zephyr/drivers/battery.h

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef CONFIG_BATTERY
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_BATTERY_H_
+#define ZEPHYR_INCLUDE_DRIVERS_BATTERY_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <zephyr/device.h>
+
+/*
+ * Sometimes we have hardware to detect battery present, sometimes we have to
+ * wait until we've been able to talk to the battery.
+ */
+enum battery_present {
+	BATTERY_PRESENCE_UNINITIALIZED = -1,
+	BATTERY_PRESENCE_NO = 0,
+	BATTERY_PRESENCE_YES = 1,
+	BATTERY_PRESENCE_UNKNOWN,
+};
+
+enum battery_disconnect_state {
+	BATTERY_DISCONNECTED = 0,
+	BATTERY_NOT_DISCONNECTED,
+	BATTERY_DISCONNECT_ERROR,
+};
+
+enum battery_property_type {
+	/* Keep these alphabetized. */
+
+	/** Runtime Dynamic Battery Parameters */
+	/**
+	 * Provide a 1 minute average of the current and voltage on the battery.
+	 * Does not check for flags or whether those values are bad readings.
+	 * See driver instance header for details on implementation and
+	 * how the average is calculated. Units in mA negative=discharging
+	 */
+	BATTERY_AVG_CURRENT,
+	BATTERY_AVG_VOLTAGE,
+	/** Whether CFET is disabled, if no CFET mask given returns false */
+	BATTERY_CHARGE_FET_DISABLED,
+	/** Battery current (mA); negative=discharging */
+	BATTERY_CURRENT,
+	/** Charge cut off from the battery */
+	BATTERY_CUT_OFF,
+	/** Cycle count in 1/100ths (number of charge/discharge cycles) */
+	BATTERY_CYCLE_COUNT,
+	/** Disconnect state of battery */
+	BATTERY_DISCONNECT_STATE,
+	/** General Error/Runtime Flags */
+	BATTERY_FLAGS,
+	/** Full Charge Capacity in mAh (might change occasionally) */
+	BATTERY_FULL_CHARGE_CAPACITY,
+	/**
+	 * Report the absolute difference between the highest and lowest cell voltage in
+	 * the battery pack, in millivolts.  On error returns '0'.
+	 */
+	BATTERY_IMBALANCE_MV,
+	/** Is the battery physically present */
+	BATTERY_PRESENT_STATE,
+	/** Remaining capacity in mAh */
+	BATTERY_REMAINING_CAPACITY,
+	/** State of charge (percent, 0-100) */
+	BATTERY_STATE_OF_CHARGE,
+	/** Temperature in 0.1 K */
+	BATTERY_TEMPERATURE,
+	/** Battery voltage (mV) */
+	BATTERY_VOLTAGE,
```

Comment:
Suggest battery_property_type to have all the commands mentioned in the smart battery spec. That way a separate API would not be needed for time_at_rate, time_to_empty and time_to_full.